src/module_3/push_exploration.md

```diff
@@ -0,0 +1,5364 @@
+print(classification_report(y_val, y_val_pred))
+
+print("Confusion Matrix:")
+cm = confusion_matrix(y_val, y_val_pred)
+print(cm)
+
+```
+
+    Classification Report:
+                  precision    recall  f1-score   support
+    
+               0       0.99      0.79      0.88    213182
+               1       0.04      0.61      0.08      3213
+    
+        accuracy                           0.78    216395
+       macro avg       0.52      0.70      0.48    216395
+    weighted avg       0.98      0.78      0.86    216395
+    
+    Confusion Matrix:
+    [[167349  45833]
+     [  1251   1962]]
+
+
+
+```python
+# Calculate probabilities for the positive class
+y_val_probs = model.predict_proba(X_val)[:, 1]
+
+# ROC Curve
+fpr_val, tpr_val, _ = roc_curve(y_val, y_val_probs)
+roc_auc_val = auc(fpr_val, tpr_val)  # Calculate AUC
+
+plt.figure(figsize=(10,5))
+
+plt.subplot(1, 2, 1)
+plt.plot(fpr_val, tpr_val, label=f'ROC (AUC = {roc_auc_val:.2f})')
+plt.plot([0, 1], [0, 1], color='navy', linestyle='--')
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('False Positive Rate')
+plt.ylabel('True Positive Rate')
+plt.title('ROC')
+plt.legend(loc="lower right")
+
+
+# Precision-Recall Curve
+precision_val, recall_val, _ = precision_recall_curve(y_val, y_val_probs)
+
+plt.subplot(1, 2, 2)
+plt.plot(recall_val, precision_val)
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('Recall')
+plt.ylabel('Precision')
+plt.title('Precision-Recall Curve')
+
+plt.tight_layout()
+plt.show()
+```
+
+
+    
+![png](push_exploration_files/push_exploration_48_0.png)
+    
+
+
+We obtain basically very similar results, so for the shake of simplicity, we will stay with the weight class that scikit provides. Curiously, both results obtain similar curves vs with no balance class for logistic regression. It may be due to the fact that, given the class 1 is so umbalanced, we do not appreciate high differences. However, we can see that in the precision in purchases (60% vs 4%, normal regression accuracy for positive is better) and recall in purchases (0% vs 62%, simple logistic regression does not find any purchase, while the later models do).
+
+Now, with these results, let's pause for a moment and analyze the numbers obtained:
+
+- Precision (True positive / true positive + false positive). Measures the accuracy of positive predictions. If the model predicts a purchase won't happen, it is correct almost always. If it predicts it will happen, only is correct 4% of the time (generating lots of false positives). Remember that current push open rate is around 5%, so it is feasible to be slightly below that. Would be useful to have the purchase rate per opened notification.
+
+- Recall or true positive rate (True positive / true positive + false negative). Measures the ability to identify all positives. If model predicts no purchase, it identifies 77% of the time. If the model predicts a purchase, it identifies 63% of the purchases
```

Comment:
same as before! the model never predicts "no purchase" it predicts a probability, so you need to combine it with a threshold